File: C++/Heap_Sort.cpp
//? The objective of this C++ Program is to implement Heap Sort in unsorted array!
#include <iostream>
using namespace std;
void heapify(int A[50], int size, int i)
{
    int largest = i;
    int left = 2 * i + 1;
    int right = 2 * i + 2;
    if (left < size and A[left] > A[largest])
        largest = left;
    if (right < size and A[right] > A[largest])
        largest = right;
    int temp = 0;
    if (largest != i)
    {
        temp = A[i];
        A[i] = A[largest];
        A[largest] = temp;
        heapify(A, size, largest);
    }
}
void Heap_Sort(int A[50], int len)
{
    int i, temp = 0;
    for (i = (len / 2) - 1; i >= 0; i--)
    {
        heapify(A, len, i);
    }
    i = len - 1;
    while (i >= 0)
    {
        temp = A[0];
        A[0] = A[i];
        A[i] = temp;
        heapify(A, i, 0);
        i--;
    }
    cout << "Your sorted array is:" << endl;
    for (i = 0; i < len; i++)
    {
        cout << A[i] << endl;
    }
}
int main()
{
    int A[50], i, j, temp = 0, l;
    cout << "Enter How Many ELements:" << endl;
    cin >> l;
    for (i = 0; i < l; i++)
    {
        cout << "Enter element" << endl;
        cin >> A[i];
    }
    cout << "Your Unsorted Array is:";
    for (i = 0; i < l; i++)
    {
        cout << A[i] << endl;
    }
    Heap_Sort(A, l);
}

File: Python/Heap_Sort.py
"""
todo:In this program, our objective is to develop a python code that applies Heap sort on a unsorted list.
? Heap Sort is similar to selection sort wehre we first find the maximum element and place the maximum element at the end.
! We repeat process for the remaining elements.
"""


def heapify(lst, size, i):
    largest = i
    left = 2*i+1
    right = 2*i+2
    if left < size and lst[left] > lst[largest]:
        largest = left
    if right < size and lst[right] > lst[largest]:
        largest = right
    temp = 0
    if largest != i:
        temp = lst[i]
        lst[i] = lst[largest]
        lst[largest] = temp
        heapify(lst, size, largest)


def heap_sort(lst, size):
    i = (size//2)-1
    while i >= 0:
        heapify(lst, size, i)
        i -= 1
    i = size-1
    temp = 0
    while i >= 0:
        temp = lst[0]
        lst[0] = lst[i]
        lst[i] = temp
        heapify(lst, i, 0)
        i -= 1


lst = []
size = int(input('How many elements:'))
i = 1
while(size != 0):
    element = int(input(f'Enter Element {i}:'))
    lst.append(element)
    i += 1
    size -= 1
print('Your Unsorted List:', lst)
heap_sort(lst, len(lst))
print('Your Sorted List[Using Heap Sort]:', lst)
print('\nHit Enter to End.......')
input()
